x/superfluid/keeper/migrate.go

```diff
@@ -1,99 +1,177 @@
+	if sharesToMigrate.IsZero() {
+		sharesToMigrate = gammSharesInLock
+	}
+
+	// Otherwise, we must ensure that the shares to migrate is less than or equal to the shares in the lock.
+	if sharesToMigrate.Amount.GT(gammSharesInLock.Amount) {
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, fmt.Errorf("shares to migrate must be less than or equal to shares in lock")
+	}
+
+	// Exit the balancer pool position.
+	exitCoins, err := k.gk.ExitPool(ctx, sender, poolIdLeaving, sharesToMigrate.Amount, sdk.NewCoins())
+	if err != nil {
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, err
+	}
+	// Defense in depth, ensuring we are returning exactly two coins.
+	if len(exitCoins) != 2 {
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, fmt.Errorf("Balancer pool must have exactly two tokens")
+	}
+
+	// Create a full range (min to max tick) concentrated liquidity position.
+	if wasSuperfluidDelegatedBeforeMigration {
+		// If the lock was previously superfluid delegated, we create a new lock and keep it locked.
+		positionId, amount0, amount1, liquidity, joinTime, concentratedLockId, err = k.clk.CreateFullRangePositionLocked(ctx, concentratedPool, sender, exitCoins, remainingLockTime)
+		if err != nil {
+			return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, err
+		}
+	} else {
+		// If the lock was unlocking, we create a new lock that is unlocking for the remaining time of the old lock.
+		positionId, amount0, amount1, liquidity, joinTime, concentratedLockId, err = k.clk.CreateFullRangePositionUnlocking(ctx, concentratedPool, sender, exitCoins, remainingLockTime)
+		if err != nil {
+			return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, err
+		}
+	}
+
+	// If the lock was previously superfluid delegated, superfluid delegate the new concentrated lock to the same validator
+	if wasSuperfluidDelegatedBeforeMigration {
+		err := k.SuperfluidDelegate(ctx, sender.String(), concentratedLockId, intermediateAccount.ValAddr)
+		if err != nil {
+			return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, err
+		}
+	}
+
+	// If the lock was superfluid undelegating at time of migration
+	if wasSuperfluidUndelegatingBeforeMigration {
+		// Create and set a new intermediary account based on the previous validator but with the new lock id and concentratedLockupDenom
+		concentratedLockupDenom := cltypes.GetConcentratedLockupDenomFromPoolId(poolIdEntering)
+		clIntermediateAccount, err := k.GetOrCreateIntermediaryAccount(ctx, concentratedLockupDenom, valAddr)
+		if err != nil {
+			return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, err
+		}
+
+		// Create a new synthetic lockup for the new intermediary account in an unlocking status
+		err = k.createSyntheticLockup(ctx, concentratedLockId, clIntermediateAccount, unlockingStatus)
+		if err != nil {
+			return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, err
+		}
```

Comment:
Can you elaborate on why we need to create this in the comment please?

Reply:
expanded on comment here https://github.com/osmosis-labs/osmosis/pull/4885/commits/9a950666cc73e17f3f4b87024e83d130e944b306